overlord/devicestate/devicemgr.go

```diff
@@ -1385,3 +1391,82 @@ func checkFDEFeatures(st *state.State, kernelInfo *snap.Info) error {
 	//       returns {"error":"hardware-unsupported"} we don't.
 	return nil
 }
+
+func (m *DeviceManager) runFDESetupHook(op string, params *boot.FdeSetupHookParams) ([]byte, error) {
+	// TODO:UC20: when this runs on refresh we need to be very careful
+	// that we never run this when the kernel is not fully configured
+	// i.e. when there are no security profiles for the hook
+
+	// state is must be locked
+	st := m.state
+
+	deviceCtx, err := DeviceCtx(st, nil, nil)
+	if err != nil {
+		return nil, fmt.Errorf("cannot get device context to run fde-setup hook: %v", err)
+	}
+	kernelInfo, err := snapstate.KernelInfo(st, deviceCtx)
+	if err != nil {
+		return nil, fmt.Errorf("cannot get kernel info to run fde-setup hook: %v", err)
+	}
+	hooksup := &hookstate.HookSetup{
+		Snap:     kernelInfo.InstanceName(),
+		Revision: kernelInfo.Revision,
+		Hook:     "fde-setup",
+		// XXX: should this be configurable somehow?
+		Timeout: 5 * time.Minute,
+	}
+	req := &hookstate.FDESetupRequest{
+		Op:      op,
+		Key:     params.Key,
+		KeyName: params.KeyName,
+		Model: map[string]string{
+			"series":     params.Model.Series(),
+			"brand-id":   params.Model.BrandID(),
+			"model":      params.Model.Model(),
+			"grade":      string(params.Model.Grade()),
+			"signkey-id": params.Model.SignKeyID(),
+		},
+		// TODO: include boot chains
+
+	}
+	contextData := map[string]interface{}{
+		"fde-setup-request": req,
+	}
+	st.Unlock()
+	defer st.Lock()
+	context, err := m.hookMgr.EphemeralRunHook(context.Background(), hooksup, contextData)
+	if err != nil {
+		return nil, fmt.Errorf("cannot run fde-setup hook for %s: %v", op, err)
```

Comment:
we probably should test this error path

Reply:
I added a test for this and also for incorrect data in fde-setup-result, I commited the later in a separate commit so that it can be reverted again, it's a bit of an internal error so maybe overkill but the test was useful for me to tweak the error a bit.